### PR TITLE
Validation visitor framework

### PR DESCRIPTION
**Pull request by gitmotte**
https://github.com/JSQLParser/JSqlParser/issues/1005

**extensible validation visitor framework**
add (package _net.sf.jsqlparser.util.validation_)
* an extensible validation visitor framework: _ValidationCapability_ and its subtypes 
  * _FeatureSetValidation_, 
  * _DatabaseMetaDataValidation_, 
  * _AllowedTypesValidation_

add (package _net.sf.jsqlparser.parser.feature_)
* FeatureConfiguration class : common way for configurations of parser-features (supports Object-values - with convenience methods for boolean getAsBoolean, String getAsString)
* Feature/FeatureSet classes : jsqlparser features which can be collected in featuresets

changes in branch https://github.com/gitmotte/JSqlParser/tree/master.validate within packages

* net.sf.jsqlparser.util.validation
* net.sf.jsqlparser.parser.feature

### CODE REVIEW

**Comment by gitmotte** [comment]
> 242

**Comment by gitmotte** [comment]
> 
> 
> [Coverage Status](https://coveralls.io/builds/33394880)
> 
> Coverage decreased (-8.4%) to 78.242% when pulling **[45514fd](https://github.com/JSQLParser/JSqlParser/commit/45514fdb677ef1d21e626f87fda7d7c1b88216e2) on gitmotte:master.validate** into **[3747f1c](https://github.com/JSQLParser/JSqlParser/commit/3747f1c00503b54dbb3fd4f7be79d28627eb52c9) on JSQLParser:master**.

ok, need to add some unit-test's ;)

**Comment by gitmotte** [comment]
@wumpz i think the next steps, bugfixes or features should be developed with new pull-requests

File: src/main/java/net/sf/jsqlparser/JSQLParserException.java [hunk 1]
```diff
@@ -12,50 +12,21 @@
...
     private static final long serialVersionUID = -1099039459759769980L;
-    private Throwable cause = null;
 
     public JSQLParserException() {
         super();
     }
 
-    public JSQLParserException(String arg0) {
-        super(arg0);
+    public JSQLParserException(String message, Throwable cause) {
```
Comment (AnEmortalKid) [inline src/main/java/net/sf/jsqlparser/JSQLParserException.java:20]: changed the class which means we should update the serial right?

**Comment by AnEmortalKid** [comment]
Should this be split in chunks?

1. 
FeatureConfiguration class : common way for configurations of parser-features (supports Object-values - with convenience methods for boolean getAsBoolean, String getAsString)
Feature/FeatureSet classes : jsqlparser features which can be collected in featuresets

2. extensible validation visitor framework

**Comment by gitmotte** [comment]
> Should this be split in chunks?

No, the validation-framework depends on Feature/FeatureSet's, features could be parser-features like switching the ability for parsing the Square-Bracket Quotation (SQLServer-Feature) on or off - those Features are configured with _FeatureConfiguration_.
The most Feature's are currently only Validation-Features, the parser supports all of them, but we like to ensure a specific syntax (i.e. for postgresql database, see _DatabaseType.POSTGRESQL_) or we like to ensure sql-scripts do not change the DDL within your database (i.e. by validate against _FeaturesAllowed.DML_)
Therefore all Feature's can be validated (by extending the Validator-Visitors), but not all Feature's allow a parser-configuration. 
And there are more validation-possibilities like 
* DatabaseMetaDataValidation (validate for names against a given database-connection)
* AllowedTypesValidation (validate the model for correctness, i.e. when using the model-builder-pattern) (not supported within the validators currently)

**Comment by gitmotte** [comment]
it's ready to merge now.

**Review by AnEmortalKid — APPROVED** [review]

**Comment by gitmotte** [comment]
@wumpz would you like to add this feature to the jsqlparser library?

**Comment by wumpz** [comment]
Sure. But til now I had no time to review this huge pr. Sorry.

-------- Ursprüngliche Nachricht --------Von: gitmotte <notifications@github.com> Datum: 24.10.20  20:17  (GMT+01:00) An: JSQLParser/JSqlParser <JSqlParser@noreply.github.com> Cc: Tobias <t.warneke@gmx.net>, Mention <mention@noreply.github.com> Betreff: Re: [JSQLParser/JSqlParser] Validation visitor framework (#1045) 
@wumpz would you like to add this feature to the jsqlparser library?

**Comment by wumpz** [comment]
Merged. Could you write some documentation to use this?

**Comment by gitmotte** [comment]
@wumpz i don't know how to add a wiki pull request - please copy the documentation to your wiki master.
https://github.com/gitmotte/JSqlParser/wiki/Examples-of-SQL-Validation

**Comment by wumpz** [comment]
Thx for this huge extension. I added your wiki page. 
👍

**Pull request merged**